src/pip/_internal/cli/req_command.py

```diff
@@ -75,6 +76,37 @@ def _get_index_urls(cls, options):
         # Return None rather than an empty list
         return index_urls or None
 
+    @classmethod
+    def _get_headers(cls, options):
+        # type: (Values) -> Optional[Dict[str, str]]
+        """
+        Return a dict of extra HTTP request headers from user-provided options.
+        """
+        # Pull header inputs
+        header_inputs = getattr(options, 'headers', [])
+        if not header_inputs:
+            return None
+
+        # Refuse to set headers when multiple index URLs are set
+        index_urls = cls._get_index_urls(options)
+        if index_urls and len(index_urls) > 1:
+            logger.warning(
+                'Refusing to set -H / --header option(s) because multiple '
+                'index URLs are configured.',
+            )
+            return None
+
+        # Parse header inputs into dict
+        headers = {}
+        for header in header_inputs:
+            match = re.match(r'^(.+?):\s*(.+)$', header.lstrip())
+            if match:
+                key, val = match.groups()
+                headers[key] = val
+            else:
+                logger.critical('Could not parse header {!r}'.format(header))
```

Comment:
Probably better to error out completely if you want to use `critical()` here. User `warning` instead if you want to carry on (but drop the malformed header).